File: advent2016/day23.rb
f=File.open(ARGV[0])
program=f.readlines().map { |line| line.strip }
def run_program(program,a=0,b=0,c=0,d=0)
  counter=0
  counter_max=program.length
  while counter<counter_max
    instruction=program[counter].split
    case instruction[0]
    when 'cpy'
      begin
        eval instruction[2]+'='+instruction[1]
      end
    when 'inc'
      eval instruction[1]+'+=1'
    when 'dec'
      eval instruction[1]+'-=1'
    when 'jnz'
      test=eval instruction[1]+'!=0'
      if test
        counter+=(eval instruction[2])-1 #-1 to offset counter increment at end of loop
      end
    when 'tgl'
      begin
        newinst=program[counter+(eval instruction[1])].split
      rescue
        counter+=1
        next
      end
      case newinst[0]
      when 'inc'
        program[counter+(eval instruction[1])]=['dec',newinst[1]].join(' ')
      when 'jnz'
        program[counter+(eval instruction[1])]=['cpy',newinst[1,2]].join(' ')
      when 'dec'
        program[counter+(eval instruction[1])]=['inc',newinst[1]].join(' ')
      when 'cpy'
        program[counter+(eval instruction[1])]=['jnz',newinst[1,2]].join(' ')
      when 'tgl'
        program[counter+(eval instruction[1])]=['inc',newinst[1]].join(' ')
      end
    end
    counter+=1
  end
  return a,b,c,d
end
a,b,c,d=run_program(program,7)
puts "Part 1: Register a = #{a}"
a,b,c,d=run_program(program,12)
puts "Part 2: Register a = #{a}"

File: advent2016/day23.py
# Author: BumpitySnook, with mod by Lains

file = open("example23.2.txt", "rb").read()
lines = file.strip().split("\n")


def is_a_num(s):
    try:
        int(s)
    except:
        return False

    return True


def get_value(regs, x):
    if is_a_num(x):
        return int(x)
    else:
        return regs[x]


def toggle(tgl_line):
    sp = tgl_line.split(" ")
    instr = sp[0]

    if instr == "inc":
        return " ".join(["dec"] + sp[1:])
    elif instr == "dec" or instr == "tgl":
        return " ".join(["inc"] + sp[1:])
    elif instr == "jnz":
        return " ".join(["cpy"] + sp[1:])
    elif instr == "cpy":
        return " ".join(["jnz"] + sp[1:])
    else:
        assert False


def assembunny_compile(lines, part2=False):
    index = 0
    regs = {"a": 7, "b": 0, "c": 0, "d": 0}

    if part2:
        regs["a"] = 12

    while True:
        if index >= len(lines):
            break

        line = lines[index]
        a, b = line.split(" ", 1)

        if a == "cpy":
            b, c = b.split(" ")
            b = get_value(regs, b)

            if c in regs:
                regs[c] = b
            else:
                print "invalid"

            index += 1

        elif a == "inc":
            if b in regs:
                #if index + 3 < len(lines) and index - 1 >= 0 and lines[index-1].startswith("cpy ") and \
                #        lines[index+1].startswith("dec") and lines[index+2].startswith("jnz") and \
                #        lines[index+3].startswith("dec") and lines[index+4].startswith("jnz"):
                #
                #    incop = b
                #    cpysrc, cpydest = lines[index-1].split(" ")[1:]
                #    dec1op = lines[index+1].split(" ")[1]
                #    jnz1cond, jnz1off = lines[index+2].split(" ")[1:]
                #    dec2op = lines[index+3].split(" ")[1]
                #    jnz2cond, jnz2off = lines[index+4].split(" ")[1:]
                #
                #    if cpydest == dec1op and dec1op == jnz1cond and \
                #        dec2op == jnz2cond and \
                #            jnz1off == "-2" and jnz2off == "-5":
                #            regs[incop] += (get_value(regs, cpysrc) * get_value(regs, dec2op))
                #            regs[dec1op] = 0
                #            regs[dec2op] = 0
                #            index += 5
                #
                #            continue
                #
                regs[b] += 1

            index += 1
        elif a == "dec":
            if b in regs:
                regs[b] -= 1

            index += 1
        elif a == "jnz":
            b, c = b.split(" ")
            b = get_value(regs, b)
            c = get_value(regs, c)

            if b != 0:
                index = index + int(c)
            else:
                index += 1
        elif a == "tgl":
            xr = b
            x = get_value(regs, xr)
            iidx = index + x

            if iidx >= 0 and iidx < len(lines):
                tgl_line = lines[iidx]
                lines[iidx] = toggle(tgl_line)
            index += 1
        else:
            assert False

    return regs["a"]


print "part 1:", assembunny_compile(lines[:])
print "part 2:", assembunny_compile(lines[:], part2=True)